armi/materials/material.py

```diff
@@ -32,6 +33,31 @@
 FAIL_ON_RANGE = False
 
 
+def parentAwareDensityRedirect(f):
+    """Wrap Material.density to warn people about potential problems.
+
+    If a Material is linked to a Component, ``Material.density`` may produce
+    different results from ``Component.density``. The component's density
+    is considered the source of truth because it incorporates changes in volume,
+    composition, and temperature in concert with the state of the reactor.
+    """
+
+    @functools.wraps(f)
+    def inner(self: "Material", *args, **kwargs) -> float:
+        if self.parent is not None:
+            runLog.warning(
+                "Calling Material.density when attached to a Component is "
+                "undesirable and can introduce subtle differences. Prefer "
+                "directly calling the parent component's density method. "
+                f"Found on {self=} attached to {self.parent=}",
+                single=True,
```

Comment:
Okay, so we have `single=True` here, so I think you want to reduce the noise on this warning. Great.

But then we are explicitly printing the ID of the parent component.

So this will print for every component someone calls `component.material.density()` on.

I think, if you're going to all the trouble to append the exact object ID, we should remove the `single=True`.  Or you could remove the ID and just do the names of the material and component class?

I don't know. But I'm not entirely convinced by the combination here.

Reply:
Hmm good points. We probably want to help discover all cases where this pattern is being used, so removing `single=True` and making a lot more noise helps that. And, at least for `Component.density` that may need to call the actual `Material.density` method, we sidestep the decorator by calling `Material.density.__wrapped__` so we shouldn't flood the logs in that case.

Removing `single=True` feels like the simplest solution. Will remove